client/asset/dcr/spv.go

```diff
@@ -0,0 +1,1009 @@
+	return w.dcrWallet.LockAccount(ctx, w.acctNum)
+}
+
+// UnlockAccount unlocks the specified account or the wallet if account is not
+// encrypted. Part of the Wallet interface.
+func (w *spvWallet) UnlockAccount(ctx context.Context, pw []byte, _ string) error {
+	return w.dcrWallet.UnlockAccount(ctx, w.acctNum, pw)
+}
+
+// SyncStatus returns the wallet's sync status.
+// Part of the Wallet interface.
+func (w *spvWallet) SyncStatus(ctx context.Context) (bool, float32, error) {
+	bestHeight := w.bestPeerInitialHeight()
+	_, h := w.dcrWallet.MainChainTip(ctx)
+	height := h
+	if h > bestHeight {
+		bestHeight = h
+	}
+
+	if h == 0 || bestHeight == 0 {
+		return false, 0, nil
+	}
+
+	return bestHeight == height, float32(height) / float32(bestHeight), nil
+}
+
+// bestPeerInitialHeight is the highest InitialHeight recorded from our peers.
+func (w *spvWallet) bestPeerInitialHeight() int32 {
+	var bestHeight int32
+	for _, p := range w.spv.GetRemotePeers() {
+		if h := p.InitialHeight(); h > bestHeight {
+			bestHeight = h
+		}
+	}
+	return bestHeight
+}
+
+// AddressPrivKey fetches the privkey for the specified address.
+// Part of the Wallet interface.
+func (w *spvWallet) AddressPrivKey(ctx context.Context, addr stdaddr.Address) (*secp256k1.PrivateKey, error) {
+	privKey, _, err := w.dcrWallet.LoadPrivateKey(ctx, addr)
+	return privKey, err
+}
+
+// cacheBlock caches a block for future use. The block has a lastAccess stamp
+// added, and will be discarded if not accessed again within 2 hours.
+func (w *spvWallet) cacheBlock(block *wire.MsgBlock) {
+	blockHash := block.BlockHash()
+	w.blockCache.Lock()
+	defer w.blockCache.Unlock()
+	cached := w.blockCache.blocks[blockHash]
+	if cached == nil {
+		cb := &cachedBlock{
+			MsgBlock:   block,
+			lastAccess: time.Now(),
+		}
+		w.blockCache.blocks[blockHash] = cb
+	} else {
+		cached.lastAccess = time.Now()
+	}
+}
+
+// cachedBlock retreives the MsgBlock from the cache, if it's been cached, else
+// nil.
+func (w *spvWallet) cachedBlock(blockHash *chainhash.Hash) *wire.MsgBlock {
+	w.blockCache.Lock()
+	defer w.blockCache.Unlock()
+	cached := w.blockCache.blocks[*blockHash]
+	if cached == nil {
+		return nil
+	}
+	cached.lastAccess = time.Now()
+	return cached.MsgBlock
+}
+
+// PeerCount returns the count of currently connected peers.
+func (w *spvWallet) PeerCount(ctx context.Context) (uint32, error) {
+	return uint32(len(w.spv.GetRemotePeers())), nil
+}
+
+// cleanBlockCache discards from the blockCache any blocks that have not been
+// accessed for > 2 hours.
+func (w *spvWallet) cleanBlockCache() {
+	w.blockCache.Lock()
+	defer w.blockCache.Unlock()
+	for blockHash, cb := range w.blockCache.blocks {
+		if time.Since(cb.lastAccess) > time.Hour*2 {
+			delete(w.blockCache.blocks, blockHash)
+		}
+	}
+}
+
+func newSpvSyncer(w *wallet.Wallet, netDir string, connectPeers []string) *spv.Syncer {
+	addr := &net.TCPAddr{IP: net.ParseIP("::1"), Port: 0}
```

Comment:
This code mirrors what dcrwallet does internally, but interestingly this address seems to be unused by dcrwallet anyway.  If you trace into p2p/peering.go, and look for where `extaddr` is used, it seems to go nowhere. It gets passed to `(*LocalPeer).newMsgVersion`, but it is unused there, using `net.Conn.LocalAddr()` instead.

Almost as if `nil` would work.  Maybe we keep it like this with ipv6 to match dcrwallet, but make a comment because it's clearly confusing multiple people.